File: Graphs.js
class Graphs {

    static getNodes(graph) {
        return graph.nodes().map(node => graph.getNodeAttributes(node));
    }

    static getRootNode(graph) {
        return Utils.getOnlyElement(Graphs.#getRootNodes(graph));
    }

    static #getRootNodes(graph) {
        return graph.nodes().filter(node => Graphs.#isRootNode(graph, node));
    }

    static #isRootNode(graph, node) {
        return graph.inNeighbors(node).length === 0;
    }
}

File: Java/src/main/java/de/keineInsektenImEssen/productsview/categoriesgraph/Graphs.java
package de.keineInsektenImEssen.productsview.categoriesgraph;

import de.keineInsektenImEssen.common.Utils;
import org.jgrapht.Graph;

import java.util.Set;
import java.util.function.Supplier;
import java.util.stream.Collectors;

public class Graphs {

    public static <V> Set<V> getRootNodes(final Graph<V, ?> graph) {
        return graph
                .vertexSet()
                .stream()
                .filter(node -> isRootNode(graph, node))
                .collect(Collectors.toSet());
    }

    private static <V> boolean isRootNode(final Graph<V, ?> graph, final V node) {
        return graph.inDegreeOf(node) == 0;
    }

    public static <V> V getRootNode(final Graph<V, ?> graph) {
        return Utils.getOnlyElement(getRootNodes(graph));
    }

    public static <V> V addAbsoluteRootNode(final Graph<V, ?> graph, final Supplier<V> absoluteRootNodeSupplier) {
        final Set<V> rootNodes = getRootNodes(graph);
        if (rootNodes.size() != 1) {
            org.jgrapht.Graphs.addOutgoingEdges(
                    graph,
                    absoluteRootNodeSupplier.get(),
                    rootNodes);
        }
        return getRootNode(graph);
    }
}
